third_party/nvfuser/csrc/type.h

```diff
@@ -72,16 +73,66 @@ enum class DataType {
   BFloat16,
   ComplexFloat,
   ComplexDouble,
-  // Vectorized types, used for reinterpret casting views
-  // TODO: add more vectorized types
-  Double_2,
-  Float_2,
   // Pointers
   SMemAddress,
   // Null
   Null
 };
 
+struct DataType;
+
+struct ArrayOf {
+  std::shared_ptr<DataType> type;
```

Comment:
Because `DataType` is a forward declaration. We need to know the definition of `DataType` to define `ArrayOf`, but we need the definition of `ArrayOf` to define `std::variant<ArrayOf, ...>`, so we need to use pointer to workaround this cyclic dependency.

Reply:
Ah, I see. I just feel this seems like a (mentally) significantly different from the current `DataType`, which is just a plain enum. 

I'm curious if the `type` field of `ArrayOf` and `PointersOf` needs to be `DataType`. Do we need to be able to express arrays of pointers, etc?

Reply:
Probably yes. In matmul kernel, there are things like
```
char *base_ptr[2] = {...};
#pragma unroll
for(i : {0, 1, 2, 3}) {
  for(j : {0, 1}) {
    char *address = base_ptr[j] + func(i)
```
where `base_ptr` is a array of pointer.

Reply:
> Probably yes. In matmul kernel, there are things like
> 
> ```c++
> char *base_ptr[2] = {...};
> #pragma unroll
> for(i : {0, 1, 2, 3}) {
>   for(j : {0, 1}) {
>     char *address = base_ptr[j] + func(i)
> ```
> 
> where `base_ptr` is a array of pointer.

I am not sure if the SASS will be different if we do it differently, but still it is good to support this case.

Reply:
Can this be just a 1D tensor of a pointer type? In other words, a tensor with a data type of `char*`?

Reply:
Tensor has some special handling in index hoisting like https://github.com/csarofeen/pytorch/blob/c6bb6318f2c97b24cdcd9bdfbbf10c73ebc9131f/third_party/nvfuser/csrc/lower_scalar_hoist.cpp#L27-L36
And these special handling are unwanted for `base_ptr`. So it would be better to use array instead of tensor.

Reply:
What's the problem with index hoisting? You don't want to apply index hoisting to `base_ptr`?

Reply:
It is the opposite. If I have
```C++
#pragma unroll
for(i : {0, 1, 2, 3}) {
  for(j : {0, 1}) {
    char *address = func1(j) + func2(i)
```
then I can not hoist the `func1(j)` part out of the `i` loop.

I want to add a scheduling primitive that allows me to cache this `func1(j)` part as `some_array[j]`, then during index hoisting, I can consider `some_array` as not having dependency on `j`, so it can be hoisted out of the `i` loop.

Reply:
And If I used `some_array` as a tensor index, then it will always be placed in the innermost for loop, and there is no index hoisting, which is not what I want.

Reply:
So since the index hoisting has the conservative assumption about tensors, it's not currently possible to hoist values held in tensors. Using arrays instead does not have the same constraint. Is this correct?

I wonder why tensors have the restriction. For example, suppose `some_array` is a 1D tensor `TensorIndex` using the loop index of `j` but not `i`. If we can find that, we can assert it doesn't depend on the loop of `i`, can't we?

Reply:
> So since the index hoisting has the conservative assumption about tensors, it's not currently possible to hoist values held in tensors. Using arrays instead does not have the same constraint. Is this correct?

Yes, correct.

> I wonder why tensors have the restriction. For example, suppose `some_array` is a 1D tensor `TensorIndex` using the loop index of `j` but not `i`. If we can find that, we can assert it doesn't depend on the loop of `i`, can't we?

We can. I think it is not a question about whether we can do it or not. It is about which way to do it is the most simple way. Supporting array of pointer is just one line of code, and I don't need to think much about it. But using tensor, I need to carefully make sure it does not mess up with things. For example, the way we treat index tensors from `index_select` and tensors from this caching mechanism is very different. For tensors from `index_select`, we need to respect its allocation and make sure that expressions depending on it is allocated after its computation. But for tensors inserted as a precomputed index cache, we want to insert allocations for it, and insert for loops to compute it. We can of course maintain a data structure to store which tensor is from fusion definition, and which is from index caching, but if we can support easily by not using tensor, why not consider that option?

Reply:
Well, I'm not sure if it's obvious which would be simpler. For supporting arrays, we need to change `DataType` to a recursive data type with smart pointers from just a plain enum. Since `DataType` is used throughout the codebase, this change seems to be quite significant. 

I don't have a particular preference here, so I'm OK with the change, but just wanted understand the reasoning of the change.

Also, I think we should have a guideline when to use arrays instead of tensors. Does it make sense to use arrays whenever during lowering? Should we limit tensors to be created as part of fusion definitions and use arrays for anything else?

Reply:
> Also, I think we should have a guideline when to use arrays instead of tensors. Does it make sense to use arrays whenever during lowering? Should we limit tensors to be created as part of fusion definitions and use arrays for anything else?

Yes, this is what I am thinking. I think this guideline also matches our programming habit (for example, we will not use an `at::Tensor` just because we want to store multiple elements). The only reason we were using tensors is that we want to use the codegen of `TensorIndex` that is already built out. But we can refactor that as well to make it support both tensor and array, and this would simplify the lowering of `ViewAsScalar`.

Ideally, I would like to see the kernel IR as close to C++ AST as possible.

> Since DataType is used throughout the codebase, this change seems to be quite significant.

I think the only affected case is `ViewAsScalar`. I don't see any other ops that were using array types. So I am not scared about potentially breaking existing use cases.

The scary part is that it enables a much larger space, that in the future, if we want to expose all dtypes to python, or if we want to serialize all dtypes/cache some intermediate results of lowering into disk, or if we heavily use things like array of array, array of pointer, etc., the work for that support could be significant. We probably would need a real parser rather than a pattern matcher to get a complete support.

> I don't have a particular preference here, so I'm OK with the change, but just wanted understand the reasoning of the change.

The reason why I prefer recursive dtypes is mostly convenience:
1. I'd like to use array vs tensor to distinguish things from lowering vs things from fusion definition
2. I'd like to use array of array so that I can do `array[i][j]` instead of `T1[i * size1 + j]`, so that I don't need to write code that computes the strided index of `T1`.

Reply:
OK, makes sense. How do we actually create an array? I see `ViewAsScalar` creates an array type, but does this PR support actually creating arrays? Can we have tests?

> Ideally, I would like to see the kernel IR as close to C++ AST as possible.

I don't think I agree. We should extend it as necessary but it should be as simple as possible.

Reply:
For example, the kernel profiler uses a `TensorView` as a profile buffer:

https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/lower_instrument.cpp#L82

Can we replace it with an array?

Reply:
> does this PR support actually creating arrays? Can we have tests?

I think it is always supported, and this PR does not change it:
```
PYTORCH_NVFUSER_DUMP="cuda_kernel" ./build/bin/nvfuser_tests --gtest_filter=*ViewAs*
```
```CUDA
__global__ void kernel1(Tensor<std::complex<float>, 1> T0, Tensor<std::complex<float>, 1> T1, Tensor<float, 2> T2, Tensor<float, 2> T7) {
  NVFUSER_DEFINE_MAGIC_ZERO
  int i45;
  i45 = ((nvfuser_index_t)threadIdx.x) * T2.stride[0];
  int i59;
  i59 = ((nvfuser_index_t)threadIdx.x) * 2;
  std::complex<float> T4[1];
  T4[0] = 0;
  T4[0]
    = T0[(((nvfuser_index_t)threadIdx.x) * T0.stride[0])]
    + T1[(((nvfuser_index_t)threadIdx.x) * T1.stride[0])];
  Array<float, 2, 1> T5[1];
  T5[0]
     = erase_type(T4[0]);
  #pragma unroll
  for(nvfuser_index_t i23 = 0; i23 < 2; ++i23) {
    int i36;
    i36 = i23 + nvfuser_zero;
    float T3[1];
    T3[0]
      = T2[(i45 + (T2.stride[1] * i36))]
      + (float) 1.00000000000000000e+00;
    float T6[1];
    T6[0] = T5[0][i23];
    T7[(i59 + i36)]
      = T3[0]
      + T6[0];
  }
  NVFUSER_UPDATE_MAGIC_ZERO
}
```

Reply:
That's a tensor of arrays, right? Isn't it possible to just create an array?

Reply:
> That's a tensor of arrays, right? Isn't it possible to just create an array?

I don't think we have the infrastructure ready to allocate an array and assign value to it right now. But I don't mind stacking another PR on top of this to demonstrate what I want to do, and how I will use it. It's already on my plate, and hopefully will provide a better justification of this.

Reply:
OK, that's fine. I'll approve the PR. Thanks for the discussion!